go/oasis-node/cmd/debug/txsource/workload/commission.go

```diff
@@ -0,0 +1,370 @@
+	for _, bound := range schedule.Bounds {
+		// Skip bounds before start bound.
+		if bound.Start <= startBound.Start {
+			continue
+		}
+		// Stop in case a bound after end is reached.
+		if bound.Start >= endEpoch {
+			break
+		}
+		boundMin := bound.RateMin.ToBigInt().Int64()
+		boundMax := bound.RateMax.ToBigInt().Int64()
+		if minBound < boundMin {
+			minBound = boundMin
+		}
+		if maxBound > boundMax {
+			maxBound = boundMax
+		}
+	}
+	// [minBound, maxBound]
+	rate := rng.Int63n(maxBound-minBound+1) + minBound
+	step := staking.CommissionRateStep{Start: startEpoch}
+	_ = step.Rate.FromInt64(rate)
+
+	return step
+}
+
+// findNextExclusiveBound finds the next Bound step that has bounds which are
+// exclusive with current bound step. Returns nil in case there is no exclusive
+// bound step.
+func findNextExclusiveBound(bounds []staking.CommissionRateBoundStep, currentBound *staking.CommissionRateBoundStep) *staking.CommissionRateBoundStep {
+	currentMin := currentBound.RateMin
+	currentMax := currentBound.RateMax
+	for _, bound := range bounds {
+		if bound.Start <= currentBound.Start {
+			continue
+		}
+		// newMin > currentMax || newMax < currentMin
+		if bound.RateMin.Cmp(&currentMax) == 1 || bound.RateMax.Cmp(&currentMin) == -1 {
+			return &bound
+		}
+		// Update bounds.
+		// newMin > currentMin
+		if bound.RateMin.Cmp(&currentMin) == 1 {
+			currentMin = bound.RateMin
+		}
+		// newMax < currentMax
+		if bound.RateMax.Cmp(&currentMax) == -1 {
+			currentMax = bound.RateMax
+		}
+	}
+	return nil
+}
+
+func (c *commission) doAmendCommissionSchedule(ctx context.Context, rng *rand.Rand, cnsc consensus.ClientBackend, stakingClient staking.Backend) error {
+	c.logger.Debug("amend commission schedule")
+
+	// Get current epoch.
+	currentEpoch, err := cnsc.GetEpoch(ctx, consensus.HeightLatest)
+	if err != nil {
+		return fmt.Errorf("GetEpoch: %w", err)
+	}
+
+	var account *staking.Account
+	account, err = stakingClient.AccountInfo(ctx, &staking.OwnerQuery{
+		Height: consensus.HeightLatest,
+		Owner:  c.account.Public(),
+	})
+	if err != nil {
+		return fmt.Errorf("stakingClient.AccountInfo %s: %w", c.account.Public(), err)
+	}
+	existingCommissionSchedule := account.Escrow.CommissionSchedule
+
+	// First epoch at which bound steps can be altered is
+	// `currentEpoch + RateBoundLead + 1`
+	// Note: Another +1 bellow since the epoch could have changed before this
+	// transaction is submitted.
+	nextAllowedBoundChangeEpoch := currentEpoch + c.rules.RateBoundLead + 1 + 1
+	// Find first epoch after `nextAllowedBoundChangeEpoch` aligned with
+	// RateChangeInterval.
+	nextAlignedBoundChangeEpoch := (((nextAllowedBoundChangeEpoch - 1) / c.rules.RateChangeInterval) + 1) * c.rules.RateChangeInterval
+
+	// Check existing bound steps. In case there are existing steps for epoch
+	// before `nextAlignedBoundChangeEpoch`, those cannot get amended and also
+	// won't be pruned yet. Therefore we need to count those to not go over the
```

Comment:
We aren't checking for steps that would be pruned. We are checking for existing steps that won't get amended by the amendment that we are making (since the earliest step in the amendment we are submitting will be at `nextAlignedBoundChangeEpoch`, existing steps for epochs smaller than it (and greater than current epoch) will be kept).

Reply:
link seems to be out of date

Reply:
> We aren't checking for steps that would be pruned.

Shouldn't we though? Because we can make the amendment longer if we take that into account

Reply:
and the comment mentions that it only counts steps that wouldn't be pruned. but the code counts steps that would be pruned too

Reply:
oh i think i know what you're saying, so the `existingCommissionSchedule` obtained by `account.Escrow.CommissionSchedule` could itself contain non pruned steps (meaning contain more than one past step with `epoch.Start<currentEpoch`) - since we actually only prune it on amendments and not automatically on every epoch.

I don't think that can actually happen at the moment since we are constantly amending schedules, the existing schedules should get pruned more or less on every epoch, but nevertheless i'll do pruning on the obtained `existingCommissionSchedule` to cover the case.

Reply:
> I don't think that can actually happen at the moment since we are constantly amending schedules, the existing schedules should get pruned more or less on every epoch

ah yeah good point